### ISSUE TITLE
Immutable lists - do we need a 'replaceInList(index, replacement)' function?

### ISSUE THREAD

**Issue by DJWalker42** [orppst/pst-api-service#3]
If an object in the database contains a list of other objects (or primitives) then the "getter" for the list returns an immutable list. This means that if we want to update an element of that list we cannot do:

`<parentObject>.getList().set(index, elementUpdate)`

where we have used "List" as a placeholder for the actual name of the list e.g. in 'TechnicalGoals' this would be 'Spectrum'. As a work around you can copy all the elements into a new list but replacing the element you wish to update with the input data, then using `<parentObject>.setList(newList)` to replace the entire list. Obviously, this is not very efficient. 

There are functions to append elements to the list i.e., `addToList(newElement)`, and to delete elements from the list i.e., `removeFromList(currentElement)`.  Should there also be a function to replace an element at a specified index e.g.,

`replaceInList(int index, Object replacementElement)` ?

**Comment by DJWalker42** [comment]
TLDR: See branch 'DJW-20-add-edit-delete-timingWindows' in pst-api-service and use the dev services. I was expecting that objects in list we want to replace retain their database ID and position in the list. Instead, the "old" item is removed and the "new", replacement item is appended to the end of list with a new database ID. Notice that we can update an individual field of an item in a list and retain the database ID but the updated item still then goes to the end of the list.

We cannot simply replace an element in a list. I have tried the following with Constraints (TimingWindows) in an Observation:
```
List<Constraint> replaceList = new ArrayList<>(observation.getConstraints());
replaceList.set(constraintIndex, replacementConstraint);

observation.setConstraints(replaceList);
```

and

```
int constraintsSize = observation.getConstraints().size();
List<Constraint> replaceList = new ArrayList<>();
for(int index = 0; index < constraintsSize; index++) {
    if (index != constraintIndex) {
        replaceList.add(observation.getConstraints().get(index));
    } else {
        replaceList.add(replacementConstraint);
    }
}

observation.setConstraints(replaceList);
```
Notice that the return value from the "replace" API call is:

```
return observation.getConstraints().get(constraintIndex);
```

which does returns the constraint data uses to replace the one at 'constraintIndex', implying thats its been replaced, but after the call returns the database actually rearranges the list. 

What seems to happen in the database after the API call has returned is equivalent to, in pseudocode:

```
remove(theOldConstraint) //all subsequent indices are shifted up by one
add(theNewConstraint) //appends the replacement item to the end of the list
```
We know this has happened because if you subsequently ask for the list of Constraints the replacement is found at the end of the list, and its database ID has been replaced with a new value. For example, lets say you have 5 constraints with consecutive database IDs starting at 1 and you try to replace the constraint found at index 2 i.e. database ID 3. After the "replace" API call returns, the database still contains 5 constraints but their database ID sequence is now 1,2,4,5,6.  

I have also tried replacing a single field, namely 'note', within a TimingWindow constraint. In this case, the database ID is retained but the newly updated TimingWindow is subsequently found at the end of the list.

For the front end this means that if we try to edit an item in a list, either by trying to replace the entire item or updating a single field in that item, that item will drop to the end of the list, and any subsequent items will shuffle up by one. Realistically, we are thus limited to simply adding or deleting items in lists until we can figure out how to replace things in lists.

**Comment by alan-stokes** [comment]
so just adding my exploration here:

1. I'm concerned how you're getting constraints not modified to change their ids. as i cant see anywhere in the database creation code which would do such things.
2. on the remove and add. id expect it to add a new id. you've removed the previous one. I think the problem is that your missing a post / put API for merging objects. such as ProposalCyclesResource line 244 for example. If we have the equiv for the objects you want to modify, id assume the ids would stay the same, and the location in the list would stay the same.
i think this is becuase the merge object under the hood might be doing something akin to a SQL UPDATE command.

**Comment by pahjbo** [comment]
I think that this became a problem early on (which we slightly glossed over by deciding to return immutable lists) because all of the JPA/Hibernate machinery expects the objects that are manipulated to be already present memory, and because we have micro services that is not true - we have to explicitly bring objects back into memory, and it is not exactly clear what the best practices are for that situation. It seems to me that there are a few options

- stick with immutable lists and then have API to replace whole list (which is current situation)
    - seems fairly inefficient in terms of database operations
-  add API to replace an individual in the list
    - could be done with a delete and then an add (though have to worry about the list ordering - this is a general problem though, sometime ordering is important sometimes not)
    - could be done with JPA merge of an object (this is just rather tedious as all fields need to be set - however this could be done as it is in the generated code)

**Comment by DJWalker42** [comment]
We may need an ["OrderBy" annotation](https://jakarta.ee/specifications/persistence/2.2/apidocs/javax/persistence/orderby) for Lists (and other Collections) in the Java code generated from the IVOA data model. 

I've found out you can replace/update an entire TimingWindow without affecting its database ID by manually using the setters and getters for each field of a TimingWindow i.e.:

```
window.setStartTime(replacementWindow.getStartTime());
window.setEndTime(replacementWindow.getEndTime());
window.setIsAvoidConstraint(replacementWindow.getIsAvoidConstraint());
window.setNote(replacementWindow.getNote());
```

hardly a scalable solution though.

**Comment by alan-stokes** [comment]
> We may need an ["OrderBy" annotation](https://jakarta.ee/specifications/persistence/2.2/apidocs/javax/persistence/orderby) for Lists (and other Collections) in the Java code generated from the IVOA data model.
> 
> I've found out you can replace/update an entire TimingWindow without affecting its database ID by manually using the setters and getters for each field of a TimingWindow i.e.:
> 
> ```
> window.setStartTime(replacementWindow.getStartTime());
> window.setEndTime(replacementWindow.getEndTime());
> window.setIsAvoidConstraint(replacementWindow.getIsAvoidConstraint());
> window.setNote(replacementWindow.getNote());
> ```
> 
> hardly a scalable solution though.

this might be me being naive, but haven't you just basically built the functionality that would be provided by the merge functionality?

**Comment by DJWalker42** [comment]
probably

**Comment by pahjbo** [comment]
I think we probably need to revisit making the returned list in the generated API immutable - it goes against many of the standard java conventions

**Comment by pahjbo** [comment]
if you pull the latest build-logic project, then a newer version of the ProposalDM will be used that returns mutable lists in the cases where there is a composition in the model. In addition there should be a .updateUsing(X x) method that is a bit like a copy constructor, but updates an exiting instance from another instance (rather than creating a new instance) - basically doing [all the laborious setting in the comment above](https://github.com/orppst/pst-api-service/issues/3#issuecomment-1814718533) - this should allow a JPA merge to be done on object from the database using the values from an object that arrives from the http API

**Comment by pahjbo** [comment]
> 
> will only update `startTime` and `endTime` that are direct children of the `TimingWindow` class. It does not update `note` and `isAvoidConstraint` that are inherited from the parent class `TimingConstraint`. Perhaps the implementation of `usingUpdate(X x)` needs to set the immediate children of `X` then call its super class version?

the implementation does directly set all the inherited members too - where in the client code is this not happening?

**Comment by pahjbo** [comment]
The problem here appears to be with what quarkus does in compile time instrumentation to make hibernate work, as doing the same test directly in the proposalDM code (where reflection is used) will pass.

**Comment by pahjbo** [comment]
will close, but it would be nice to find out why "joined" JPA inheritance was not apparently working properly
